src/gradient_descent.c

```diff
@@ -0,0 +1,380 @@
+/// function
+/// @param f Function we want to transfor to a 1 dimension function
+/// @param initialPoint Initial point to base the movement of
+/// @param direction Direction in which to move
+/// @param n Number of dimensions of the original function
+/// @return Structure containing the bracket for the minimum
+static Bracket bracketMinimum(f1dimension func, function f,
+                              gradient_real initialPoint[],
+                              gradient_real direction[], int n) {
+  // Set a, b and c at arbitrary initial values
+  Bracket bracket = {0.0, 1.0, 0.0};
+
+  // Evaluate function at initial points a and b
+  gradient_real fa = (*func)(f, bracket.a, initialPoint, direction, n);
+  gradient_real fb = (*func)(f, bracket.b, initialPoint, direction, n);
+
+  // We want to go downhill from a to b to bracket the minimum
+  // so if fb > fa we need to swap them
+  if (fb > fa) {
+    gradient_real temp = bracket.a;
+    bracket.a = bracket.b;
+    bracket.b = temp;
+
+    temp = fa;
+    fa = fb;
+    fb = fa;
+  }
+
+  // Initial guess for c
+  bracket.c = (bracket.b) + GOLD * (bracket.b - bracket.a);
+  gradient_real fc = (*func)(f, bracket.c, initialPoint, direction, n);
+
+  // Loop until we bracket the minimum
+  while (fb > fc) {
+    // Compute u the inflexion point of the parabola defined by
+    // points a, b and c
+    gradient_real r = (bracket.b - bracket.a) * (fb - fc);
+    gradient_real q = (bracket.b - bracket.c) * (fb - fa);
+    gradient_real diff = q - r;
+    if (diff == 0) {
+      diff = EPS; // Prevent division by 0
+    }
+    gradient_real inflexion = (bracket.b) - ((bracket.b - bracket.c) * q -
+                                             (bracket.b - bracket.a) * r) /
+                                                (2.0 * diff);
+
+    // Value of function at inflexion point
+    gradient_real fInflexion;
+
+    // Set a maximum distance for where we will look for a minimum
+    gradient_real ulim = (bracket.b) + GLIMIT * (bracket.c - bracket.b);
+
+    // Check if u is between b and c
+    if ((bracket.b - inflexion) * (inflexion - bracket.c) > 0.0) {
+
+      // Compute function value at u
+      fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+
+      // Check if we have a minimum between b and c
+      if (fInflexion < fc) {
+        bracket.a = bracket.b;
+        bracket.b = inflexion;
+        return bracket;
+      }
+
+      // Check if we have a minimum between a and u
+      else if (fInflexion > fb) {
+        bracket.c = inflexion;
+        return bracket;
+      }
+
+      // Our parabolic interpolation was not useful
+      // Magnify u and proceed to the next iteration
+      inflexion = (bracket.c) + GOLD * (bracket.c - bracket.b);
+      fInflexion = (*func)(f, inflexion, initialPoint, direction, n);
+    }
+
+    // Check if u is between c and the limit
+    else if ((bracket.c - inflexion) * (inflexion - ulim) > 0.0) {
```

Comment:
In the book numerical recipes in C, the condition is >= 0, why did we change it ?